source/hpack.c

```diff
@@ -782,79 +804,90 @@ int aws_hpack_decode_integer(
-        return aws_raise_error(AWS_ERROR_SHORT_BUFFER);
-    }
+    const size_t output_len_backup = output->len;
+
+    /* Determine length of encoded string (and whether or not to use huffman) */
+    uint8_t use_huffman;
+    size_t str_length;
+    switch (context->huffman_mode) {
+        case AWS_HPACK_HUFFMAN_NEVER:
+            use_huffman = 0;
+            str_length = to_encode.len;
+            break;
+
+        case AWS_HPACK_HUFFMAN_ALWAYS:
+            use_huffman = 1;
+            str_length = aws_huffman_get_encoded_length(&context->encoder, to_encode);
+            break;
+
+        case AWS_HPACK_HUFFMAN_SMALLEST:
+            str_length = aws_huffman_get_encoded_length(&context->encoder, to_encode);
+            if (str_length < to_encode.len) {
+                use_huffman = 1;
+            } else {
+                str_length = to_encode.len;
+                use_huffman = 0;
+            }
+            break;
 
-    struct aws_byte_cursor to_encode_backup = *to_encode;
+        default:
+            aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
+            goto error;
+    }
 
-    /* Write the use_huffman bit */
-    output->buffer[output->len] = huffman_encode << 7;
+    /*
+     * String literals are encoded like so (RFC-7541 5.2):
+     * H is whether or not data is huffman-encoded.
+     *
+     *   0   1   2   3   4   5   6   7
+     * +---+---+---+---+---+---+---+---+
+     * | H |    String Length (7+)     |
+     * +---+---------------------------+
+     * |  String Data (Length octets)  |
+     * +-------------------------------+
+     */
 
-    /* Write the header */
-    size_t encoded_length;
-    if (huffman_encode) {
-        encoded_length = aws_huffman_get_encoded_length(&context->encoder, *to_encode);
-    } else {
-        encoded_length = to_encode->len;
-    }
-    if (aws_hpack_encode_integer(encoded_length, 7, output)) {
+    /* Encode string length */
+    uint8_t starting_bits = use_huffman << 7;
+    if (aws_hpack_encode_integer(str_length, starting_bits, 7, output)) {
+        HPACK_LOGF(ERROR, context, "Error encoding HPACK integer: %s", aws_error_name(aws_last_error()));
         goto error;
     }
 
-    if (huffman_encode) {
-        aws_huffman_encoder_reset(&context->encoder);
-        int result = aws_huffman_encode(&context->encoder, to_encode, output);
-        if (result) {
-            goto error;
-        }
-    } else {
-        bool result = aws_byte_buf_write_from_whole_cursor(output, *to_encode);
-        if (!result) {
-            aws_raise_error(AWS_ERROR_SHORT_BUFFER);
-            goto error;
+    /* Encode string data */
+    if (str_length > 0) {
+        if (use_huffman) {
+            /* Huffman encoder doesn't grow buffer, so we ensure it's big enough here */
+            if (s_ensure_space(output, str_length)) {
+                goto error;
+            }
+
+            if (aws_huffman_encode(&context->encoder, &to_encode, output)) {
```

Comment:
Didn't you just add a huffman_encode that grows dynamically? Am I mis-remembering?

Reply:
I modified huffman DECODE to grow dynamically.
Needed that because you don't know how much space the compressed data will expand to unless you run through it all twice, but it's streaming in so you don't have access to it all up front, so it's basically impossible.

For encoding, we have access to all the data before we encode it, so we can just figure out exactly how much space it will need ahead of time.

I actually started modifying the huffman encoder to grow dynamically, then realized I didn't need to and reverted it